Objects/dictobject.c

```diff
@@ -1650,6 +1650,69 @@ PyDict_SetItem(PyObject *op, PyObject *key, PyObject *value)
     return insertdict(mp, key, hash, value);
 }
 
+/* Initialize dict from keywords tuple and values array.
+ *
+ * - Dict must be empty before first call, presized, and combined.
+ * - keys must be distinct.
+ */
+static int
+dict_init_from_items(PyObject *op, PyObject *keys, PyObject * const *values)
+{
+    assert(PyDict_Check(op));
+    assert(PyTuple_Check(keys));
+
+    PyDictObject *mp = (PyDictObject *)op;
+    assert(!_PyDict_HasSplitTable(mp));
+
+    Py_ssize_t size = PyTuple_GET_SIZE(keys);
+    assert(mp->ma_used == 0);
+    assert(mp->ma_keys->dk_usable >= size);
+
+    PyDictKeyEntry *ep = &DK_ENTRIES(mp->ma_keys)[0];
+
+    int all_unicode = 1;
+
+    for (Py_ssize_t i = 0; i < size; i++) {
+        PyObject *key = PyTuple_GET_ITEM(keys, i);
+        Py_INCREF(key);
+        if (!PyUnicode_CheckExact(key)) {
+            all_unicode = 0;
+        }
+
+        Py_hash_t hash;
+        if (!PyUnicode_CheckExact(key) ||
```

Comment:
PyUnicode_CheckExact is done above, you can store it in a var.

Reply:
PyUnicode_CheckExact is simple function. No need to use new flag variable to cache it.